### ISSUE TITLE
extractTranscriptSeqs fails reconstructing transcripts on minus strand

### ISSUE THREAD

**Issue by sghignone** [Bioconductor/GenomicFeatures#12]
When working on my favourite genome, I have noticed that extractTranscriptSeqs assembles cds in wrong order, when trying to reconstruct transcripts located on tthe minus strand. 

In order to demonstrate this issue, I have a test script developed on random coordinates on chr2 of the Human genome.

`#load libraries
library("GenomicFeatures")
library("Biostrings")
library("BSgenome.Hsapiens.UCSC.hg19")
#load genome
genome <- BSgenome.Hsapiens.UCSC.hg19
#build GRanges object: let's say there are 2 genes (named match), the first with just one 'exon' (match_part), 
#and the second with 3 exons, two of which are overlapping. 
( gr <- GRanges(Rle(rep("chr2", 6)),
              IRanges(start=c(11874,11874,14874,14874,15374,15674), end=c(12874,12874,15874,15474,15574,15874)),
              Rle(strand(c("-")), c(6)),
              type=c("match","match_part","match","match_part","match_part","match_part"),
              target=c("A","A","B","B","B","B"),
              parentFix=c("a-A","a-A","b-B","b-B","b-B","b-B")
    ))

#Select MATCHES, create GRangesList object, run merge -- non expecting any merge --, 
#extract the genomic sequence based on Iranges
( gr.m <- gr[gr$type == "match"] )
( gr.m.grl = split(gr.m, as.factor(gr.m$parentFix)) )
( mergers.m.minus <- IRanges::reduce(gr.m.grl, with.revmap=TRUE) )
( cds.m.minus <- extractTranscriptSeqs(genome, mergers.m.minus) )
#Select MATCH_PARTS, create GRangesList object, run merge -- expecting overlap between feature 1 and 2 --, 
#extract the genomic sequence based on Iranges
( gr.mp <- gr[gr$type == "match_part"] )
( gr.mp.grl = split(gr.mp, as.factor(gr.mp$parentFix)) )
( mergers.mp.minus <- IRanges::reduce(gr.mp.grl, with.revmap=TRUE) )
( cds.mp.minus <- extractTranscriptSeqs(genome, mergers.mp.minus) )
#COMMENT ON THE RESULT: The first transcript (one cds) has been correctly extracted, while the cds' of the second transcript have been mounted in the wrong order. 
#To fix this issue the following line helps: sort coordiantes in decreasing order.
( cds.mp.minus <- extractTranscriptSeqs(genome, sort(mergers.mp.minus,decreasing=T)) )
`
As stated in documentation:
[..] For a transcript located on the minus strand, the exons will typically (but not necessarily) be ordered by descending position on the reference genome. 

...but it's not working at the moment.

**Comment by hpages** [comment]
Hi,

When the transcripts argument of `extractTranscriptSeqs()` is a GRangesList object (like in your case), each list element of this object must be a GRanges object representing the exons of a given transcript. As explained in the man page, these exons must be ordered by ascending **rank**, that is, by ascending position **in the transcript** (when going in the 5' to 3' direction). This means that the first exon in `transcripts[[i]]` is considered to be the first exon (rank 1) in transcript `i`, the 2nd exon in `transcripts[[i]]` the 2nd exon (rank 2) in transcript `i`, etc... `extractTranscriptSeqs()` doesn't look at how the exons are positioned with respect to one another **on the reference genome**.

For example, let's take a close look at the 2nd transcript in your `mergers.mp.minus` object. This transcript has 2 exons:
```
mergers.mp.minus[[2]]
# GRanges object with 2 ranges and 1 metadata column:
#       seqnames      ranges strand |        revmap
#          <Rle>   <IRanges>  <Rle> | <IntegerList>
#   [1]     chr2 14874-15574      - |           1,2
#   [2]     chr2 15674-15874      - |             3
#   -------
#   seqinfo: 1 sequence from an unspecified genome; no seqlengths
```
`extractTranscriptSeqs()` considers that this transcript is made of exons `mergers.mp.minus[[2]][1]` and `mergers.mp.minus[[2]][2]`. So it will extract the DNA sequences corresponding to these 2 exons and concatenate them **in that order**. Now if we look at the locations of these 2 exons on the reference genome, we notice that the 1st exon is located downstream of the 2nd exon, when going in the 5' to 3' direction, which is not the typical situation for a real-world transcript. This is what the following comment in the man page is about:

    For a transcript located on the minus strand, the exons will typically (but not
    necessarily) be ordered by descending position on the reference genome.

This comment is not about what  `extractTranscriptSeqs()` expects, it's about how exons in a real-world transcript are typically arranged.

If you give `extractTranscriptSeqs()` an artificial/unrealistic transcript where the above is not true, it's not `extractTranscriptSeqs()` problem. The reason `extractTranscriptSeqs()` takes it is that there are actually some rare real-world transcripts where the above is not true. 

In other words it's your responsibility to make sure that the GRangesList object you supply to `extractTranscriptSeqs()` contains the exons ordered **by rank**. Note that this order is also the order in which translation will happen, whatever the original order of the exons on the reference genome is.

Hope this helps,
H.

**Comment by hpages** [comment]
@sghignone , I'm closing this issue since the above seems to have answered your question.
